lib/AST/Module.cpp

```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
   return nullptr;
 }
 
+//===----------------------------------------------------------------------===//
+// Cross-Import Overlays
+//===----------------------------------------------------------------------===//
+
+namespace swift {
+/// Represents a file containing information about cross-module overlays.
+class OverlayFile {
+  /// If not empty, contains the name of the cross-import file that must be
+  /// loaded; \c overlayModuleNames is not valid and must be empty.
+  ///
+  /// If empty, \c overlayModuleNames is valid and will contain the list of
+  /// module names unless loading the file failed.
+  StringRef filePath;
+
+  /// The list of module names; empty if loading failed.
+  llvm::TinyPtrVector<Identifier> overlayModuleNames;
+
+  /// Actually loads the overlay module name list. This should mutate
+  /// \c overlayModuleNames, but not \c filePath.
+  void loadOverlayModuleNames(ModuleDecl *M, SourceLoc diagLoc,
+                              Identifier bystandingModule);
+
+public:
+  // Only allocate in ASTContexts.
+  void *operator new(size_t bytes) = delete;
+  void *operator new(size_t bytes, const ASTContext &ctx,
+                     unsigned alignment = alignof(OverlayFile)) {
+    return ctx.Allocate(bytes, alignment);
+  }
+
+  OverlayFile(StringRef filePath)
+    : filePath(filePath), overlayModuleNames() { }
+
+  /// Returns the list of additional modules that should be imported if both
+  /// the primary and secondary modules have been imported. This may load a
+  /// file; if so, it will diagnose any errors itself and arrange for the file
+  /// to not be loaded again.
+  ///
+  /// The result can be empty, either because of an error or because the file
+  /// didn't contain any overlay module names.
```

Comment:
I don't understand this bit. If the file didn't contain any overlay module names, is that not an error?

Reply:
I don't diagnose it as an error—it's just an array with zero entries. Maybe I should.

Reply:
That seems like a mistake on the part of whoever wrote the overlay. Since we will see the problem in a client, not the overlay author... I'm feeling a bit divided on if we should give a hard error or not. It seems unlikely that someone would ship a totally broken overlay with no entries so maybe the hard error is fine, and we want to detect that sooner rather than later in case that happens. A warning is definitely not what we want because the client cannot suppress the warning without going in and deleting/modifying the file.

Reply:
Suppose someone, for instance, ran their swiftoverlay files through the C preprocessor (or gyb, or whatever) so they could use `#if` to remove overlays related to unreleased features:

```
---
version: 1
modules:
#if !HIDE_RAY_GUNS
- name: _MyModule_OtherModule_RayGuns
#endif
#if !HIDE_FLYING_CARS
- name: _MyModule_OtherModule_FlyingCars
#endif
```

This could totally result in a swiftoverlay file with no modules listed. It'd probably be better if they erased the file, but it's not really doing any harm.

Reply:
That's a good point! I'm convinced.